pkg/cache/scheduler_callback_test.go

```diff
@@ -0,0 +1,653 @@
+				TerminationType: si.TerminationType_STOPPED_BY_RM,
+			},
+		},
+	})
+	assert.NilError(t, err, "error updating allocation")
+	assert.Assert(t, !context.schedulerCache.IsAssumedPod(taskUID1))
+	err = utils.WaitForCondition(deleteCalled.Load, 10*time.Millisecond, 500*time.Millisecond)
+	assert.Error(t, err, "timeout waiting for condition") // pod is not expected to be deleted
+}
+
+func TestUpdateAllocation_AskReleased(t *testing.T) {
+	callback, context := initCallbackTest(t, false, true)
+	defer dispatcher.UnregisterAllEventHandlers()
+	defer dispatcher.Stop()
+	app := context.getApplication(appID)
+	app.sm.SetState(ApplicationStates().Running)
+	var deleteCalled atomic.Bool
+	context.apiProvider.(*client.MockedAPIProvider).MockDeleteFn(func(pod *v1.Pod) error { //nolint:errcheck
+		deleteCalled.Store(true)
+		return nil
+	})
+
+	err := callback.UpdateAllocation(&si.AllocationResponse{
+		ReleasedAsks: []*si.AllocationAskRelease{
+			{
+				ApplicationID:   appID,
+				AllocationKey:   taskUID1,
+				TerminationType: si.TerminationType_TIMEOUT,
+			},
+		},
+	})
+	assert.NilError(t, err, "error updating allocation")
+	assert.Assert(t, !context.schedulerCache.IsAssumedPod(taskUID1))
+	err = utils.WaitForCondition(deleteCalled.Load, 10*time.Millisecond, time.Second)
+	assert.NilError(t, err, "pod has not been deleted")
+}
+
+func TestUpdateApplication_Accepted(t *testing.T) {
+	callback, context := initCallbackTest(t, false, false)
+	defer dispatcher.UnregisterAllEventHandlers()
+	defer dispatcher.Stop()
+	app := context.getApplication(appID)
+	app.sm.SetState(ApplicationStates().Submitted)
+
+	err := callback.UpdateApplication(&si.ApplicationResponse{
+		Accepted: []*si.AcceptedApplication{
+			{
+				ApplicationID: appID,
+			},
+		},
+	})
+	assert.NilError(t, err, "error updating application")
+	err = utils.WaitForCondition(func() bool {
+		return app.sm.Current() == ApplicationStates().Accepted
+	}, 10*time.Millisecond, time.Second)
+	assert.NilError(t, err, "application has not transitioned to Accepted state")
+}
+
+func TestUpdateApplication_Rejected(t *testing.T) {
+	callback, context := initCallbackTest(t, false, false)
+	defer dispatcher.UnregisterAllEventHandlers()
+	defer dispatcher.Stop()
+	app := context.getApplication(appID)
+	app.sm.SetState(ApplicationStates().Submitted)
+	NewPlaceholderManager(context.apiProvider.GetAPIs())
+	recorder := k8sEvents.NewFakeRecorder(1024)
+	events.SetRecorder(recorder)
+
+	err := callback.UpdateApplication(&si.ApplicationResponse{
+		Rejected: []*si.RejectedApplication{
+			{
+				ApplicationID: appID,
+				Reason:        "test failure",
+			},
+		},
+	})
+	assert.NilError(t, err, "error updating application")
+	err = utils.WaitForCondition(func() bool {
+		return app.sm.Current() == ApplicationStates().Failed
+	}, 10*time.Millisecond, time.Second)
+	assert.NilError(t, err, "application has not transitioned to Failed state")
+	assert.Equal(t, 1, len(recorder.Events), "no K8s event received")
```

Comment:
not sure why we verify the dispatch only for `reject` event?

Reply:
Updated both `TestUpdateApplication_Failing` and `TestUpdateApplication_Failed` as those also generate events.